### ISSUE TITLE
Change Request: Export default config

### ISSUE THREAD

**Issue by voxpelli** [eslint/eslint#18619]
### ESLint version

9.5.0

### What problem do you want to solve?

Eg. the `@eslint/config-inspector` needs to duplicate the ESLint default config as well as configuration resolution logic and file matching logic, leading to potential mismatches – see eg https://github.com/eslint/config-inspector/pull/66 and https://github.com/eslint/config-inspector/pull/59

I can imagine that similar tools, such as my [`compare-eslint-configs`](https://github.com/voxpelli/compare-eslint-configs) will have similar challenges when they start to support flat configs

### What do you think is the correct solution?

I would want these to be exposed:

* [`calculateConfigArray`](https://github.com/eslint/eslint/blob/21d3766c3f4efd981d3cc294c2c82c8014815e6e/lib/eslint/eslint.js#L369)
* [`findFiles`](https://github.com/eslint/eslint/blob/21d3766c3f4efd981d3cc294c2c82c8014815e6e/lib/eslint/eslint-helpers.js#L515)

The `calculateConfigArray` enables one to get a properly configured `FlatConfigArray` for a project, including default values, solving https://github.com/eslint/config-inspector/pull/66 for good.

The `findFiles` then enables one to properly find all matching files for that config, avoiding most of https://github.com/eslint/config-inspector/pull/59 and https://github.com/eslint/config-inspector/pull/63 – where the correct files are matched and the correct files are ignored.

### Participation

- [X] I am willing to submit a pull request for this change.

### Additional comments

Since these are API:s for tools like `@eslint/config-inspector` and `compare-eslint-configs` they are not core to the functionality of ESLint itself and hence I think `unsupported-api.js` is the correct place for them.

**Comment by nzakas** [comment]
We can't just expose these APIs as they were not designed for public use and doing so would make it more difficult for us to make changes to these APIs in the future.

Can you explain more about what config-inspector does and what it needs?

The solution is probably more along the lines of defining new APIs rather than just exposing private ones.

**Comment by voxpelli** [comment]
> Can you explain more about what config-inspector does and what it needs?

Two things:

1. When it resolves the config it needs to know the default config which ESLint add and while that can be copied (as it did in https://github.com/eslint/config-inspector/pull/66) it would be more self-explanatory and less error prone if it imported the very default config that the relevant version of ESLint adds.

2. When resolving the files from a config set there are a lot of things to consider, in how different file patterns are handled and in how ignores are applied. The major way to get this right across multiple implementations would be to share the very same implementation. Both https://github.com/eslint/config-inspector/issues/59 and https://github.com/eslint/config-inspector/issues/63 shows shortcomings in its current lookup and in https://github.com/eslint/config-inspector/issues/59 @mdjermanovic points at the specific inner workings of `getConfigWithStatus()` in `@eslint/config-array` and re-implementing all of that logic seems error prone and duplicated effort.

Hence why I suggested [calculateConfigArray](https://github.com/eslint/eslint/blob/21d3766c3f4efd981d3cc294c2c82c8014815e6e/lib/eslint/eslint.js#L369) and [findFiles](https://github.com/eslint/eslint/blob/21d3766c3f4efd981d3cc294c2c82c8014815e6e/lib/eslint/eslint-helpers.js#L515) – as the latter need the former and the former solves adding the correct default config.

And since `calculateConfigArray()` returns a `FlatConfigArray` any additions to `@eslint/config-array` would become available through that one as well, making introspection of configs more powerful without necessarily exposing much of the inner workings of `eslint`.

Eg: To solve the UX for https://github.com/eslint/config-inspector/issues/59 `config-inspector` would need to be able to inspect whether a specific config applies directly through en explicit match or indirectly through a universal pattern or lack of pattern, something which `@eslint/config-array` already reasons about in `getConfigWithStatus()` and could probably be made to expose the the logic for.

Either `config-inspector` constructs its own `ConfigArray` to enable such things or it makes use of an ESLint specific factory like `calculateConfigArray()`.

I think we can agree that working on and resolving config arrays is something that should involve `@eslint/config-array` in some way? Either explicitly by constructing ones own `ConfigArray` or implicitly through eg. `calculateConfigArray()`? Currently `config-inspector` does not use `@eslint/config-array`

**Comment by nzakas** [comment]
> I think we can agree that working on and resolving config arrays is something that should involve `@eslint/config-array` in some way?

Yes. I'd say the best place to start is just to use `@eslint/config-array` directly. The file matching logic is all encapsulated there and you'll get the same results as in ESLint. It seems like that solves most of problems?

Then I think the other thing to address is the default configuration from ESLint. This is tricky because the default configuration may change with each release. I'm actually trying to minimize the default configuration now to simplify things. I need to think on this a bit more.

**Comment by voxpelli** [comment]
> Yes. I'd say the best place to start is just to use `@eslint/config-array` directly. The file matching logic is all encapsulated there and you'll get the same results as in ESLint. It seems like that solves most of problems?

The one thing lacking might be some of the logic from within [`globSearch()`](https://github.com/eslint/eslint/blob/d2d06f7a70d9b96b125ecf2de8951bea549db4da/lib/eslint/eslint-helpers.js#L217-L382), so I decided to make a quick simplified version of it: https://github.com/voxpelli/config-array-find-files

Such a function could maybe be useful to upstream to `@eslint/config-array` itself eventually, and to rewrite `globSearch()` to base itself upon the `configArrayFindFiles()` rather than directly on `@nodelib/fs.walk`

**Comment by voxpelli** [comment]
The `flatConfigSchema` would kind of also be needed to load an ESLint flat config into `@eslint/config-array`, as else it complains of eg. `Unexpected key "languageOptions" found.`

https://github.com/eslint/eslint/blob/21d3766c3f4efd981d3cc294c2c82c8014815e6e/lib/config/flat-config-schema.js#L548-L566

**Comment by voxpelli** [comment]
The config-inspector now uses `@eslint/config-array` at its core, using a minimal schema: https://github.com/eslint/config-inspector/blob/ca6070e86bd6a5aa6f11cf1d9fc9bd0d9e956e97/src/configs.ts#L212-L243

**Comment by nzakas** [comment]
Thinking this through, I can't think of a reason why it wouldn't be safe to export the default internal config that ESLint uses. Maybe export it from `eslint/configs/default`, like:

```js
const defaultConfig = require("eslint/configs/default");
```

As long as people realize that the default internal config may change between releases, I can't think of a downside to doing this.

@eslint/eslint-tsc what do you think?

**Comment by fasttime** [comment]
> Thinking this through, I can't think of a reason why it wouldn't be safe to export the default internal config that ESLint uses. Maybe export it from `eslint/configs/default`, like:
> 
> ```js
> const defaultConfig = require("eslint/configs/default");
> ```
> 
> As long as people realize that the default internal config may change between releases, I can't think of a downside to doing this.
> 
> @eslint/eslint-tsc what do you think?

Makes sense to me, if that's helpful for the use cases of `@eslint/config-inspector` and `compare-eslint-configs`.

**Comment by mdjermanovic** [comment]
Since it may change between releases, it might be better to add it in `unsupported-api.js` (`eslint/use-at-your-own-risk`)?

Default config also provides access to core rules (for example, `defaultConfig[0].plugins["@"].rules["no-unused-vars"]`) which we didn't want to provide in supported APIs.

**Comment by voxpelli** [comment]
Maybe one can skip the first of the configs and only expose these? To avoid exposing rules etc https://github.com/eslint/eslint/blob/e2b16e2b72606162dce3d804bc80186b6c5ec0f9/lib/config/default-config.js#L56-L74

**Comment by mdjermanovic** [comment]
Okay, marking as accepted to export the default config as described in https://github.com/eslint/eslint/issues/18619#issuecomment-2187154709.

**Comment by nzakas** [comment]
Another option I just thought of: what if we made this a static property of the `ESLint` class?

```js
const { ESLint } = require("eslint");

const configs = [ ...ESLint.defaultConfig ];
```

Or we could even make it a symbol property if we didn't want to make it obvious:

```js
const { ESLint, ESLintSymbol } = require("eslint");

const configs = [ ...ESLint[ESLintSymbol.defaultConfig] ];
```

**Comment by nzakas** [comment]
@eslint/eslint-tsc what do you think of [this suggestion](https://github.com/eslint/eslint/issues/18619#issuecomment-2189529900)?

**Comment by mdjermanovic** [comment]
> Or we could even make it a symbol property if we didn't want to make it obvious:
> 
> ```js
> const { ESLint, ESLintSymbol } = require("eslint");
> 
> const configs = [ ...ESLint[ESLintSymbol.defaultConfig] ];
> ```

Sounds good to me.

**Comment by fasttime** [comment]
> Another option I just thought of: what if we made this a static property of the `ESLint` class?
> 
> ```js
> const { ESLint } = require("eslint");
> 
> const configs = [ ...ESLint.defaultConfig ];
> ```
> 
> Or we could even make it a symbol property if we didn't want to make it obvious:
> 
> ```js
> const { ESLint, ESLintSymbol } = require("eslint");
> 
> const configs = [ ...ESLint[ESLintSymbol.defaultConfig] ];
> ```

I'd say both options make sense, only the usage would be different. The first option looks simpler for users, and for maintainers also, but that's a detail.

**Comment by nzakas** [comment]
The first option is if we want to make it obvious to people that it's there to use.

The second option is if we want to make it clear that this isn't intended to be used all that often but is present.

I think the first option makes it easier to feature test the existence of the new property.

**Comment by voxpelli** [comment]
One could possibly do the first but name it something like `.referenceDefaultConfig`, to make it clear that its to be used as a reference, not as a config

**Comment by nzakas** [comment]
@voxpelli I'm not sure I understand the distinction you're trying to make. What is the difference between a reference and a config?

**Comment by voxpelli** [comment]
@nzakas My thought was that there's a difference between _using_ a config and _referencing_ a config, and if one wants to avoid confusion where people start using this config in their own configs, then being explicit that it's there for tools to reference, not by end users to use in their own configs

But that's probably me overthinking it, let's keep it simple

**Comment by nzakas** [comment]
Ah gotcha. So the consequence of people using it in their own config is minimal, it'll just duplicate information that's already there and be normalized away.

**Comment by nzakas** [comment]
Making it readonly is a good idea. I think we should also freeze the value so it can't be changed.

**Comment by mdjermanovic** [comment]
Looks like we have all the details defined now. @voxpelli would you like to submit a PR to implement this?

**Comment by voxpelli** [comment]
Sure, I can do that, I will try to get to it soon 👍
